pkg/executor/join/hash_join_v2.go

```diff
@@ -18,32 +18,46 @@ import (
 	"bytes"
 	"context"
 	"math"
+	"math/rand"
 	"runtime/trace"
 	"strconv"
 	"sync"
 	"sync/atomic"
 	"time"
 
+	"github.com/pingcap/errors"
 	"github.com/pingcap/failpoint"
 	"github.com/pingcap/tidb/pkg/executor/internal/exec"
 	"github.com/pingcap/tidb/pkg/expression"
 	"github.com/pingcap/tidb/pkg/parser/mysql"
 	"github.com/pingcap/tidb/pkg/planner/core/operator/logicalop"
+	"github.com/pingcap/tidb/pkg/sessionctx/variable"
 	"github.com/pingcap/tidb/pkg/types"
 	"github.com/pingcap/tidb/pkg/util"
 	"github.com/pingcap/tidb/pkg/util/channel"
 	"github.com/pingcap/tidb/pkg/util/chunk"
 	"github.com/pingcap/tidb/pkg/util/disk"
 	"github.com/pingcap/tidb/pkg/util/execdetails"
 	"github.com/pingcap/tidb/pkg/util/memory"
+	"github.com/pingcap/tidb/pkg/util/serialization"
 )
 
+const minimalHashTableLen = 32
+
 var (
 	_ exec.Executor = &HashJoinV2Exec{}
 	// enableHashJoinV2 is a variable used only in test
 	enableHashJoinV2 = atomic.Bool{}
 )
 
+func getHashTableLength(table *rowTable) uint64 {
+	return max(nextPowerOfTwo(table.validKeyCount()), uint64(minimalHashTableLen))
+}
+
+func getHashTableMemoryUsage(hashTableLength uint64) int64 {
+	return int64(hashTableLength) * serialization.UnsafePointerLen
```

Comment:
> And why define `getHashTableLength` and `getHashTableMemoryUsage` in this file instead of `hash_table_v2.go`?

I will move them to `hash_table_v2.go`